daemon/cmd/fqdn.go

```diff
@@ -158,6 +160,47 @@ func (d *Daemon) updateSelectorCacheFQDNs(ctx context.Context, selectors map[pol
 	return d.endpointManager.UpdatePolicyMaps(ctx, notifyWg)
 }
 
+// deletedEndpointNameCollector collects the FQDN entries from deleted endpoints'
+// DNSCaches which can then be used to GC the FQDN entries in cases where endpoints
+// matching FQDN policies still remain, but do not contact these names.
+type deletedEndpointNameCollector struct {
+	lock.Mutex
+	names map[string]struct{}
+}
+
+// EndpointCreated implements endpointmanager.Subscriber.
+func (*deletedEndpointNameCollector) EndpointCreated(ep *endpoint.Endpoint) {
+}
+
+// EndpointDeleted implements endpointmanager.Subscriber.
+func (dc *deletedEndpointNameCollector) EndpointDeleted(ep *endpoint.Endpoint, conf endpoint.DeleteConfig) {
+	dc.Lock()
```

Comment:
Great question to ask. I did not think about it. Let's check (to some reasonable manual search depth). Reverse call stack, mutexes in `{}`:
```
DNSZombieMappings.Upsert
  DNSCache.GC {DNSCache.RWMutex}
    dns-garbage-collect-job {}

DNSZombieMappings.Names
   deletedEndpointNameCollector.EndpointDeleted {}
     endpointManager.removeEndpoint {endpointManager.mutex}
       endpointManager.RemoveEndpoint {}
         Daemon.deleteEndpointQuiet {}
           (many callers, no locks in sight)

DNSZombieMappings.GC
   dns-garbage-collect-job {}
   bootstrapFQDN {}
   deleteDNSLookups {}
     deleteFqdnCacheHandler {}

DNSZombieMappings.MarkAlive
  Endpoint.MarkDNSCTEntry {}
    ctmap.gc.Enable {}

DNSZombieMappings.SetCTGCTime
  Endpoint.MarkCTGCTime {}
    ctmap.gc.Enable {}

DNSZombieMappings.ForceExpire
  deleteDNSLookups {}

DNSZombieMappings.ForceExpireByNameIP
  notifyOnDNSMsg {getMutexesForResponseIPs}
    DNSProxy.ServeDNS {}

DNSZombieMappings.DumpAlive
  extractDNSLookups {}
    getFqdnCacheHandler {}

DNSZombieMappings.(Unm|M)arshalJSON
  (can't figure out from references)    
```

```
DNSCache.DisableCleanupTrack
  bootstrapFQDN {}

DNSCache.Update
  dns-garbage-collector-job {}
  bootstrapFQDN {}
  notifyOnDNSMsg {getMutexesForResponseIPs}
  deleteDNSLookups {}

DNSCache.GC
  dns-garbage-collector-job {}
  bootstrapFQDN {}

DNSCache.UpdateFromCache
  bootstrapFQDN {}
  deleteDNSLookups {}
  
DNSCache.ReplaceFromCacheByNames
  dns-garbage-collector-job {}
  
DNSCache.Lookup + LookupByRegexp
  NameManager.MapSelectorsToIPsLocked {}
    NameManager.UpdateGenerateDNS {NameManager.RWMutex}
      Daemon.notifyOnDNSMsg {getMutexesForResponseIPs}
 
DNSCache.LookupIP
  Daemon.GetNamesOf {}
    (hubble parsers)
       LocalObserverServer.Start: {}

DNSCache.ForceExpire
  deleteDNSLookups {}

DNSCache.Names
   deletedEndpointNameCollector.EndpointDeleted {}
     endpointManager.removeEndpoint {endpointManager.mutex}
       endpointManager.RemoveEndpoint {}
         Daemon.deleteEndpointQuiet {}
           (many callers, no locks in sight)

DNSCache.Dump
  extractDNSLookups {}
    getFqdn*Handler {}
 ```

Based on this quick survey, it looks like there are no paths back from DNSCache/DNSZombieMappings to acquire endpoint manager locks, nor does it appear that the locks from DNSCache/DNSZombieMappings are leaking (e.g. no callbacks passed into them).

Reply:
> DNSZombieMappings.(Unm|M)arshalJSON

I think this is called by the endpoint base64 stuff, which means we also hold the EP lock while doing it. Specifically the `Marshal`ling I think happens regularly via `syncHeaderFile`. 

```
ep.base64() {}
  ep.writeInformationalComment {}
    ep.writeHeaderFile {}
      ep.syncHeaderFile {}
        ...
      ep.runPreCompilationSteps() { EP lock }
        ep.regenerateBPF()
          ...
```

But afaik the endpoint manager will `ep.Delete()` the EP before calling its subscribers, so the controller running the period EP header file sync should be stopped by the time we reach this, hence no deadlock possible, unless I missed something.

Reply:
Even if it did happen, holding the endpoint lock should be fine?

Reply:
It may be, but historically the EP lock has been involved in many deadlocks. I'm not sure we have the guarantee that we never hold an EP lock when acquiring the EP manager lock. for example. But yeah, it's getting a bit academic, I think this is fine :)